server/src/services/metadata.service.ts

```diff
@@ -478,6 +488,146 @@ export class MetadataService implements OnEvents {
+  }
+
+  private toPixels(region: MetadataRegion, unit: string): MetadataRegion {
+    const pixelRegion: MetadataRegion = region;
+    if (unit.toLowerCase() == 'normalized') {
+      return {
+        imageWidth: region.imageWidth,
+        imageHeight: region.imageHeight,
+        x1: Math.floor(region.x1 * region.imageWidth),
+        x2: Math.floor(region.x2 * region.imageWidth),
+        y1: Math.floor(region.y1 * region.imageHeight),
+        y2: Math.floor(region.y2 * region.imageHeight),
+      };
+    }
+    return pixelRegion;
+  }
+
+  private async applyTaggedFaces(asset: AssetEntity, tags: ImmichTags) {
+    this.logger.debug(`Extracting face metadata from (${asset.id})`);
+
+    if (!tags.RegionInfo) {
+      return true;
+    }
+
+    // Stores all face regions
+    const faceSet = new Set<MetadataFaceResult>();
+
+    // Stores all face names
+    const nameSet = new Set<string>();
+
+    // Stores AssetFace entities
+    const discoveredFaces: Partial<AssetFaceEntity>[] = [];
+
+    for (const region of tags.RegionInfo?.RegionList) {
+      const faceResultName = region.Name || 'unknown';
+      const faceResult: MetadataFaceResult = {
+        Name: faceResultName,
+        Type: region.Type,
+        Region: this.toPixels(
+          {
+            imageWidth: tags.RegionInfo?.AppliedToDimensions?.W,
+            imageHeight: tags.RegionInfo?.AppliedToDimensions?.H,
+            x1: region.Area.X - region.Area.W / 2,
+            y1: region.Area.Y - region.Area.H / 2,
+            x2: region.Area.X + region.Area.W / 2,
+            y2: region.Area.Y + region.Area.H / 2,
+          },
+          region.Area.Unit,
+        ),
+      };
+      faceSet.add(faceResult);
+      nameSet.add(faceResultName);
+    }
+    this.logger.debug(`${faceSet.size} faces detected in ${asset.originalPath}`);
+
+    if (nameSet.size === 0) {
+      return true;
+    }
+
+    // Find all person entries matching the face names we have
+    const matches = await this.personRepository.getManyByName(asset.ownerId, [...nameSet], { withHidden: true });
+
+    // Identify which names are not in our inventory yet
+    const matchesNames = new Set(matches.map((m) => m.name.toLowerCase()));
+    const missing = new Set([...nameSet].filter((item) => !matchesNames.has(item.toLowerCase())));
+
+    if (missing.size > 0) {
+      this.logger.debug(`Creating missing persons: ${[...missing]}`);
+    }
+
+    const newPersons = await this.personRepository.createMany(
+      [...missing].map((name) => ({ ownerId: asset.ownerId, name: name })),
+    );
+    if (!newPersons) {
+      this.logger.error(`Something went wrong creating persons`);
+    }
+
+    // All persons, initially found matching face names in this asset and newly created ones
+    const persons = [...matches, ...newPersons];
+
+    // Loop to build all required faceAssets elements
+    for (const face of faceSet) {
+      const facePerson = persons.find((fp) => fp.name.toLowerCase() == face.Name);
```

Comment:
I'm reviewing on my phone so take this with a grain of salt, but it seems like you use sets and still end up iterating to look up values. Sounds like a case for a map?

Reply:
Yup, agree Map is a good fit for these

Reply:
https://github.com/immich-app/immich/commit/f2856d96c389ace7176348805104b47d39c9de0d#diff-61222a9e424956861d68b088bbb08a464f1fa5d8a8c8fa7a3cfa116ae1f8170f